static/src/javascripts/projects/common/utils/steady-page.js

```diff
@@ -0,0 +1,188 @@
+                return height + readHeight(insertion.container);
+            }, 0);
+        });
+
+        function elementIsAbove(el) {
+            var parentElArray = [];
+            var parentEl = el.container;
+            var viewportAdjustment = Math.max(document.documentElement.clientHeight, window.innerHeight || 0) / 2;
+
+            // Push the container's parents into an array so we can calculate
+            // the position of the elements all the way until the body
+            while (parentEl) {
+                parentElArray.push(parentEl);
+                parentEl = parentEl.offsetParent;
+            }
+
+            // Get the top position of the element + half of the viewport (to account for an element loading
+            // just below the top of the viewport)
+            var elTopPos = parentElArray.reduce(function(topPos, parentEl){
+                // This will loop up the parents until the body taking into account all the positions
+                // so that if an element is equal to the scrollY position then this will return 0
+                return topPos + parentEl.offsetTop + parentEl.clientTop;
+            }, 0) - viewportAdjustment;
+
+            // If the distance of the element from the top of the page minus the height of the
+            // element we are measuring is less than the scroll position then we know that it will
+            // push the page down when loading in
+            return el.container.offsetHeight > -1 &&
+                window.scrollY > 0 &&
+                (elTopPos - el.container.offsetHeight < window.scrollY);
+        }
+
+        function readHeight(el) {
+            var style = getComputedStyle(el);
+            var height = el.offsetHeight + parseInt(style.marginTop) + parseInt(style.marginBottom);
+            return isNaN(height) ? 0 : height;
+        }
+    }
+
+
+    /**
+     * Given a batch, call all of the callbacks on the insertion object
+     *
+     * @param  {Array} batch
+     */
+    function insertElements (batch) {
+        return fastdom.write(function () {
+            batch.forEach(function (insertion) {
+                insertion.cb();
+            });
+        });
+    }
+
+    /**
+     * Process the insertion operation:
+     *
+     *   1. Calculate the original height of the container
+     *   2. Apply the insertion functions for all inserted elements
+     *   3. Calculate the new height of the container
+     *   4. Adjust the scroll position to account for the new container height
+     *
+     * @param  {Number} state.scrollY The original scroll position
+     */
+    function go(state) {
+        running = true;
+
+        var batch = [];
+        var batchHeightsBeforeInsert;
+
+        promise = fastdom.read(function(){
+                while (!q.empty()) {
+                    // Take the current queue items and add them to the batch array
+                    var insertion = q.dequeue();
```

Comment:
`dequeue` as an operation is a different thing to a double ended queue (I think some people call them `deque`s). 
there's plenty of examples of it out there, e.g. [scala](http://www.scala-lang.org/api/2.7.7/scala/collection/mutable/Queue.html#dequeue%28%29), [wikipedia](https://en.wikipedia.org/wiki/Queue_(abstract_data_type))